Skiller simulator extension to lookup skill execs from mongodb

TarikViehmann opened the pull request:
### General
This PR introduces the plugin `skiller-simulator-lookup`, which similar to the `skiller-simulation-navgraph` plugin gives a more elaborate provider of execution skill times. The idea is to recycle skill executions of past runs and to feed them the skiller-simulator. The `mongodb_log` plugin (specifically the blackboard logger) may be used to record skill executions in real scenarios which then form the baseline for skiller-simulator-based executions. The plugin checks if records for a given skill are present. If so, it provides a skill duration by performing a random lookup from all matching records.

### Features
1. **Lookup Exception List:** In order to play nicely together with the nagraph-based estimator it is possible to configure the lookup-based estimator to try all skills that are not covered by `skiller-simulator-navgraph`. To do so one may specify the same list of skills as the navgraph-estimator uses and additionally enable the `try-by-default` flag that causes the lookup of all skills except the ones specified in the skill list.
2. **Argument Sensitivity:** It may be necessary to distinguish skills not only by their name but also by the given args. This is possible by setting the config value `match-args` to `true`.
3. **Speed Scaling:**  Perform faster-than-realtime simulations by manipulating the scaling factor `speed`.
4. **Failure Simulation:** It is also possible to simulate execution failures by replaying skill execution outcomes as well. This  may be enabled via the `match-failures` option

### Interactions with `mongodb_log`
As mentioned earlier the `mongodb_log` plugin may be used to record regular skill-executions. It offers the possibility to log all blackboard messages from any interface. In order to use this data it has to be processed first in order to calculate skill execution times from the different messages. To help with that the python script `mongodb_bb_connector`is provided.

### Open Questions
I decided to make this a draft PR first as i feel there might be some need for changes. In addition to any input that comes to your mind, i want to briefly go over the stuff that i was unsure about:
 - **Q1:** Which mongodb instance should contain the lookup database? In the first draft i added a new instance on 27019. I did want to have the data somewhere, where i do not clear it by accident. But having a separate instance just for those statistics is probably unnecessary. So maybe it just should be the default 27017 instance or even put the database inside `robot-memory-local`?
 - **Q2:** Should i provide a `c++` tool instead of a python script to get the data? I am more fluent with the `pymongo` driver than `mongocxx` and therefore used python for my experiments. However, i later saw that other plugins provide `c++` tools and that with fawkes argument parsing functions it would not be too hard to make a tool in similar fashion.
 - **Q3:** This PR contains a change in `mongodb_log` that logs the thread id along with the blackboard messages. Reason being that it could happen that multiple robots could execute the same skill with the same arguments at the same time. In order to distinguish between the different skill calls i added the thread id which feels kinda random, but is there a better way to solve this problem?
 - **Q4:** Should i split up this PR? Maybe it makes sense to have changes on the existing plugins separate from this one.

morxa commented:
First of all, great work, this looks pretty nice!

> 
>     1. **Lookup Exception List:** In order to play nicely together with the nagraph-based estimator it is possible to configure the lookup-based estimator to try all skills that are not covered by `skiller-simulator-navgraph`. To do so one may specify the same list of skills as the navgraph-estimator uses and additionally enable the `try-by-default` flag that causes the lookup of all skills except the ones specified in the skill list.

I find the naming sub-optimal. This is basically treating the list of skills as blacklist, correct? Can't we just have a blacklist instead?

I also think we should re-consider the config of the different providers. It already started getting messy with #197, but it gets worse with this PR. The config layout should be the same for each estimator. That being said, this is out of scope for this PR, so I'm even fine if the resulting config of this plugin is not perfect.

> 
>     2. **Argument Sensitivity:** It may be necessary to distinguish skills not only by their name but also by the given args. This is possible by setting the config value `match-args` to `true`.

Does that work for all providers or only for the lookup provider?


> 
>     * **Q1:** Which mongodb instance should contain the lookup database? In the first draft i added a new instance on 27019. I did want to have the data somewhere, where i do not clear it by accident. But having a separate instance just for those statistics is probably unnecessary. So maybe it just should be the default 27017 instance or even put the database inside `robot-memory-local`?

We should definitely use one of the existing ones. Using `robot-memory-local` would mean that each robot has its own database. It may be easier to parse the results if we collect them centrally, i.e., on the off-field machine (27017 in simulation). The disadvantage is that this creates additional network traffic, but this is probably negligible. Another advantage of the central instance is that its data is never pruned, in contrast to the other instances.

> 
>     * **Q2:** Should i provide a `c++` tool instead of a python script to get the data? I am more fluent with the `pymongo` driver than `mongocxx` and therefore used python for my experiments. However, i later saw that other plugins provide `c++` tools and that with fawkes argument parsing functions it would not be too hard to make a tool in similar fashion.

The tool is only needed off-line, correct? That is, you run it once after collecting data? If that's true, a python script is fine, as it does not need to be integrated.

> 
>     * **Q3:** This PR contains a change in `mongodb_log` that logs the thread id along with the blackboard messages. Reason being that it could happen that multiple robots could execute the same skill with the same arguments at the same time. In order to distinguish between the different skill calls i added the thread id which feels kinda random, but is there a better way to solve this problem?

What about the robot's name? It's currently an agent-specific config value, but we could change it to be a general config parameter.

> 
>     * **Q4:** Should i split up this PR? Maybe it makes sense to have changes on the existing plugins separate from this one.

I think it's fine.

TarikViehmann commented:
Thanks for the input!

> > **Lookup Exception List:** In order to play nicely together with the nagraph-based estimator it is possible to configure the lookup-based estimator to try all skills that are not covered by `skiller-simulator-navgraph`. To do so one may specify the same list of skills as the navgraph-estimator uses and additionally enable the `try-by-default` flag that causes the lookup of all skills except the ones specified in the skill list.
> 
> I find the naming sub-optimal. This is basically treating the list of skills as blacklist, correct? Can't we just have a blacklist instead?

Sure, i did not do it initially as i wanted to provide the default whitelist option as used by the navgraph estimator. Having a blacklist as well means one has to consider the corner cases, such as empty whitelist and blacklist = nothing allowed (same behavior as empty skills list in navgraph estimator), empty whitelist nonempty blacklist = everything except blacklist allowed.

> I also think we should re-consider the config of the different providers. It already started getting messy with #197, but it gets worse with this PR. The config layout should be the same for each estimator. That being said, this is out of scope for this PR, so I'm even fine if the resulting config of this plugin is not perfect.

Yea i fully agree that this is starting to get out of control. We could make the loading of common config values part of the aspect manager, such that `speed`, `skills`, `match-args` and similar configs that should be considered by any estimator are loaded upon provider registration. Then we could also centralize the whole blacklist whitelist behaviour instead of doing it for all estimators seperately.

> > **Argument Sensitivity:** It may be necessary to distinguish skills not only by their name but also by the given args. This is possible by setting the config value `match-args` to `true`.
> 
> Does that work for all providers or only for the lookup provider?

It only works for the lookup provider right now, it is relatively straightforward to do it for the other providers but it would require to retrieve the skill args from the config as well (currently i only get them from the database). So users would need to write the whole skill string into the configs, which then has to be parsed. To be really usable one should also add ways to only partially specify skill args.
> > **Q1:** Which mongodb instance should contain the lookup database? In the first draft i added a new instance on 27019. I did want to have the data somewhere, where i do not clear it by accident. But having a separate instance just for those statistics is probably unnecessary. So maybe it just should be the default 27017 instance or even put the database inside `robot-memory-local`?
> 
> We should definitely use one of the existing ones. Using `robot-memory-local` would mean that each robot has its own database. It may be easier to parse the results if we collect them centrally, i.e., on the off-field machine (27017 in simulation). The disadvantage is that this creates additional network traffic, but this is probably negligible. Another advantage of the central instance is that its data is never pruned, in contrast to the other instances.

Yea lets just use the off-field machine for that.

> > **Q3:** This PR contains a change in `mongodb_log` that logs the thread id along with the blackboard messages. Reason being that it could happen that multiple robots could execute the same skill with the same arguments at the same time. In order to distinguish between the different skill calls i added the thread id which feels kinda random, but is there a better way to solve this problem?
> 
> What about the robot's name? It's currently an agent-specific config value, but we could change it to be a general config parameter.

That would make the most sense to me.

morxa commented:
@TarikViehmann when you update this PR, can you also fix the issues Codacy is complaining about? Thanks!

TarikViehmann commented:
I think content-wise the work is done now. I will wait for #226 to be merged, then rebase my stuff and change this PR from draft to regular.

Short update list:
 - some bundlng of config loading in execution time estimator parent class
 - centralized and argument-sensitive black/whitelisting considerations
 - no separate mongodb instance anymore, instead configurable instance to one of the existing ones

TarikViehmann commented:
I think i am done with the rebasing now, i also prepared a followup branch in fawkes-robotino (https://github.com/carologistics/fawkes-robotino/tree/tviehmann/estimator-config-update) which may be helpful if you want to test the new stuff (just adapting the configs to the changes).

morxa commented on src/plugins/execution-time-estimator-navgraph/navgraph_estimator.cpp:
```diff
@@ -35,45 +35,47 @@ namespace fawkes {
 
 bool
-NavGraphEstimator::can_execute(const Skill &skill) const
+NavGraphEstimator::can_provide_exec_time(const Skill &skill)
```
As mentioned above, this can be kept `const` if the mutex is declared `mutable`

TarikViehmann replied:
Done in d100bffa49be6efa274fc1464d7613dedf209dda

TarikViehmann commented on src/plugins/execution-time-estimator-lookup/lookup_estimator.h:
```diff
@@ -0,0 +1,75 @@
+	constexpr static char skill_name_field_[] = "name";
+	constexpr static char duration_field_[]   = "duration";
+
+	fawkes::Mutex     mutex_;
```
Nice, TIL i learned about `mutable`! I think `get_execution_time` should not be const regardless as the calculation of `get_execution_time` may generally impact the later call of `execute`. The lookup estimator does this by saving the skill execution result of the entry that provides the exec time.

TarikViehmann commented on src/plugins/execution-time-estimator-lookup/lookup_estimator.cpp:
```diff
@@ -0,0 +1,192 @@
+{
+	error_feedback = error_;
+	return outcome_;
+}
```
Adressed in https://github.com/fawkesrobotics/fawkes/pull/218/commits/c4a305e83aa315133ba43d89dfac5e0f77e5675f, adapted skiller simulator with structured binding in https://github.com/fawkesrobotics/fawkes/pull/218/commits/43797b0f88d5771c8b029b2a08b3aadbd4c64f1f.

TarikViehmann commented on src/plugins/execution-time-estimator-lookup/lookup_estimator.cpp:
```diff
@@ -0,0 +1,192 @@
+		}
+		bsoncxx::stdx::optional<bsoncxx::document::value> found_entry =
+		  mongodb_client_lookup_->database(database_)[collection_].find_one(query.view());
+		return bool(found_entry);
```
Adressed in https://github.com/fawkesrobotics/fawkes/pull/218/commits/f36dfd56c1a7dbf39d027c458f0137532ec00f94.

TarikViehmann commented on src/plugins/execution-time-estimator-lookup/lookup_estimator.cpp:
```diff
@@ -0,0 +1,192 @@
+			return;
+		} catch (fawkes::Exception &) {
+			using namespace std::chrono_literals;
+			logger_->log_info(logger_name_, ("Waiting for mongodb " + instance_ + " instance").c_str());
```
Adressed in https://github.com/fawkesrobotics/fawkes/pull/218/commits/f36dfd56c1a7dbf39d027c458f0137532ec00f94.

TarikViehmann commented on src/plugins/execution-time-estimator-lookup/lookup_estimator.cpp:
```diff
@@ -0,0 +1,192 @@
+	logger_->log_info(logger_name_,
+	                  ("Using instance " + instance_ + " database " + database_ + " collection "
+	                   + collection_)
+	                    .c_str());
```
Adressed in https://github.com/fawkesrobotics/fawkes/pull/218/commits/f36dfd56c1a7dbf39d027c458f0137532ec00f94.

TarikViehmann commented on src/plugins/execution-time-estimator-lookup/lookup_estimator.cpp:
```diff
@@ -0,0 +1,192 @@
+		}
+	}
+	if (!get_property(include_failures_)) {
+		query.append(kvp("outcome", (int)SkillerInterface::SkillStatusEnum::S_FINAL));
```
Adressed in https://github.com/fawkesrobotics/fawkes/pull/218/commits/f36dfd56c1a7dbf39d027c458f0137532ec00f94.

TarikViehmann commented on src/plugins/execution-time-estimator-lookup/lookup_estimator.cpp:
```diff
@@ -0,0 +1,192 @@
+	MutexLocker lock(mutex_);
+	try {
+		if (get_property(include_failures_)) {
+			query.append(kvp("outcome", (int)SkillerInterface::SkillStatusEnum::S_FINAL));
```
Adressed in https://github.com/fawkesrobotics/fawkes/pull/218/commits/f36dfd56c1a7dbf39d027c458f0137532ec00f94.

TarikViehmann commented on src/plugins/execution-time-estimator-lookup/lookup_estimator.cpp:
```diff
@@ -0,0 +1,192 @@
+		float res = 0.f;
+		switch (doc[duration_field_].get_value().type()) {
+		case bsoncxx::type::k_double: res = float(doc[duration_field_].get_double().value); break;
+		case bsoncxx::type::k_int32: res = float(doc[duration_field_].get_int32().value); break;
```
Adressed in https://github.com/fawkesrobotics/fawkes/pull/218/commits/f36dfd56c1a7dbf39d027c458f0137532ec00f94.

TarikViehmann commented on src/libs/execution_time_estimator/execution_time_estimator.cpp:
```diff
@@ -42,17 +44,98 @@ using Skill = ExecutionTimeEstimator::Skill;
+  whitelist_(get_skills_from_config(cfg_prefix_ + "whitelist")),
+  blacklist_(get_skills_from_config(cfg_prefix_ + "blacklist"))
+{
+}
```
Adressed in https://github.com/fawkesrobotics/fawkes/pull/218/commits/1d328a7db48418c4cbe3af2d81e6cb22ad0e7da0.

TarikViehmann commented on src/libs/execution_time_estimator/execution_time_estimator.cpp:
```diff
@@ -96,4 +194,115 @@ Skill::parse_args(const std::string &args)
+ExecutionTimeEstimator::get_skills_from_config(const std::string &path) const
+{
+	const int                                     ID       = 0;
+	const int                                     PROPERTY = 1;
```
Adressed in https://github.com/fawkesrobotics/fawkes/pull/218/commits/0700edd831db9519168cf543d803731da2cae994 and https://github.com/fawkesrobotics/fawkes/pull/218/commits/8ecbf1073c2f6ddca1f6623fd9c58c9bfe5d4bfe.

TarikViehmann commented on src/libs/execution_time_estimator/execution_time_estimator.cpp:
```diff
@@ -96,4 +194,115 @@ Skill::parse_args(const std::string &args)
+		std::vector<std::string> skill_property =
+		  str_split(std::string(it->path()).substr(path.size()));
+		if (skill_property.size() != 2) {
+			break;
```
Indeed, good catch! fixed in https://github.com/fawkesrobotics/fawkes/pull/218/commits/edc998ce6b7eead0585b6ab2f29cd7047c4ceb0b.

TarikViehmann commented on src/libs/execution_time_estimator/execution_time_estimator.cpp:
```diff
@@ -42,17 +44,98 @@ using Skill = ExecutionTimeEstimator::Skill;
- * @fn bool ExecutionTimeEstimator::can_execute(const Skill &skill) const
- * Check if this estimator can give an estimate for the given
+ * @fn bool ExecutionTimeEstimator::can_provide_exec_time(const Skill &skill)
+ * Check if this estimator can give an estimate for a given
```
Found it weird as well, but copy pasted that patern. Fixed in https://github.com/fawkesrobotics/fawkes/pull/218/commits/bb1176fbe94156f85dcbcf7b997fd42178accd3e, https://github.com/fawkesrobotics/fawkes/pull/218/commits/c4a305e83aa315133ba43d89dfac5e0f77e5675f and https://github.com/fawkesrobotics/fawkes/pull/218/commits/3e2b6874e4e60f0b18e555dd5133f556d08f4282.

TarikViehmann commented on src/libs/execution_time_estimator/execution_time_estimator.cpp:
```diff
@@ -42,17 +44,98 @@ using Skill = ExecutionTimeEstimator::Skill;
+ *
+ * @fn std::map<std::string, Skill> ExecutionTimeEstimator::get_skills_from_config(const std::string &path) const
+ * Load skill descriptions from a yaml config. The skills are represented via
+ * suffixes /<description-id>/name and /<description-id>/args following the
```
Fixed in https://github.com/fawkesrobotics/fawkes/pull/218/commits/bb1176fbe94156f85dcbcf7b997fd42178accd3e, https://github.com/fawkesrobotics/fawkes/pull/218/commits/c4a305e83aa315133ba43d89dfac5e0f77e5675f and https://github.com/fawkesrobotics/fawkes/pull/218/commits/3e2b6874e4e60f0b18e555dd5133f556d08f4282.

TarikViehmann commented on src/libs/execution_time_estimator/execution_time_estimator.cpp:
```diff
@@ -96,4 +194,115 @@ Skill::parse_args(const std::string &args)
+			skill_strings[skill_property[ID]] += str_join(it->get_strings(), ',');
+		} else if (skill_property[PROPERTY] == "name") {
+			skill_strings[skill_property[ID]] =
+			  it->get_string() + "{" + skill_strings[skill_property[ID]];
```
No this works, the `name` case prepends the name and `{`, the `args` case appends the args, hence the skill string will be built correctly. The only case that will cause a crash is if `args` is specified but `name` is not. It will cause: `13:50:21.736368 FawkesMainThread: [EXCEPTION] Unexpected skill string: 'shelf=.*}'
`, which i guess is fine.

TarikViehmann commented on src/libs/execution_time_estimator/execution_time_estimator.cpp:
```diff
@@ -96,4 +194,115 @@ Skill::parse_args(const std::string &args)
+		default_value = default_val;
+	}
+	const int                                     ID             = 0;
+	const int                                     PROPERTY       = 1;
```
Adressed in 8ecbf1073c2f6ddca1f6623fd9c58c9bfe5d4bfe and 0700edd831db9519168cf543d803731da2cae994.

TarikViehmann commented on src/plugins/execution-time-estimator-lookup/lookup_estimator.cpp:
```diff
@@ -0,0 +1,192 @@
+		for (const auto &skill_arg : active_whitelist_entry_->second.skill_args) {
+			query.append(kvp("args." + skill_arg.first, skill_arg.second));
+		}
+	}
```
Yea i agree, done in 99756727467947648c8fc959edf1a396f505b313.

TarikViehmann commented on src/plugins/execution-time-estimator-lookup/lookup_estimator.h:
```diff
@@ -0,0 +1,75 @@
+
+private:
+	void init();
+
```
There is no need for init, fixed in 174b9a7c02269d72661a13b702c29c7b8756f9c7

TarikViehmann commented on src/libs/execution_time_estimator/execution_time_estimator.cpp:
```diff
@@ -96,4 +194,115 @@ Skill::parse_args(const std::string &args)
+		} else if constexpr (std::is_same<T, bool>()) {
+			default_value = config->get_bool((path + property).c_str());
+		} else {
+			throw Exception("Unsupported type");
```
I chose a different route of `static_assert` because the conditions of `enable_if` would have to be specified in source and header file. My solution is implemented in 2d3b8ebd4160d7b1ad034f22f695ab22c8ecfe96

TarikViehmann commented on src/plugins/execution-time-estimator-lookup/lookup_estimator.cpp:
```diff
@@ -0,0 +1,192 @@
+	  config_->get_uint_or_default((mongo_cfg_prefix + "startup-grace-period").c_str(), 10);
+	logger_->log_info(logger_name_, ("Connect to mongodb " + instance_ + " instance").c_str());
+	unsigned int startup_tries = 0;
+	for (; startup_tries < startup_grace_period * 2; ++startup_tries) {
```
Sounds reasonable, i wrapped the creation of the mongocxx client in `mongodb_client_config.cpp`with the waiting logic and shifted the configuration into the `mongodb.yaml`. Done in https://github.com/fawkesrobotics/fawkes/pull/218/commits/000bfe05d2935bb9f7e60ae1b968768937886b51 and https://github.com/fawkesrobotics/fawkes/pull/218/commits/d0329993154e1f10c5db933ec7457378ff0d4b13. Also, the logic can be dropped in `robot-memory` due to this: https://github.com/fawkesrobotics/fawkes/pull/218/commits/6f7d1b147c89ab9e82f7f63baf2e6069a2bbbd7d

TarikViehmann commented on src/libs/execution_time_estimator/execution_time_estimator.cpp:
```diff
@@ -42,17 +44,98 @@ using Skill = ExecutionTimeEstimator::Skill;
  *
- * @fn void ExecutionTimeEstimator::execute(const Skill &skill) const
+ * @fn bool ExecutionTimeEstimator::can_execute(const Skill &skill)
+ * Check if this estimator is both allowed and able to give an estimate for a given
```
Fixed in bb1176f, c4a305e and 3e2b687.

TarikViehmann commented on src/plugins/Makefile:
```diff
@@ -53,6 +54,7 @@ gazebo: robotino
 skiller: navgraph
 skiller-simulator: skiller execution-time-estimator
 execution-time-estimator-navgraph: execution-time-estimator
+sexecution-time-estimator-lookup: execution-time-estimator mongodb
```
Oh yea, screwed that one up doing rebasing, fixed now in ba7f38e87ee187c4b20a156300f07c2a00cade4a.

TarikViehmann commented:
> I have a number of (smaller) remarks. Also, as a general comment, the python scripts should be installed/symlinked to `bin/`.

Thanks for the great feedback! I think i adressed it all now, symlinking to bin got fixed in https://github.com/fawkesrobotics/fawkes/pull/218/commits/abcbf04730196722d1c5c012a1a167c23be68fe2.

morxa commented on cfg/conf.d/execution-time-estimator.yaml:
```diff
@@ -2,17 +2,75 @@
+        name: say
+        exec-time: 12
+      # matches, if "hello world" has to be said
+      01-say-hello-world:
+        name: say
+        args: ["text=hello world"]
+        exec-time: 17
```
Did you test this? I would have expected that `hello world` needs to be quoted, it's a string after all.

TarikViehmann replied:
> Is there a reason why you don't put the script into `etc/scripts`? This would also take care of installing it automatically.

I oriented myself on the plugin assisting tool of the `mongodb` (there it is indeed a c++ tool so it makes sense to only compile it if the plugin itself is built) plugin, so i initially placed it in the the plugin folder. Remembering our handling of scripts in fawkes-robotino i thought about placing it in `etc/scripts`, however there was no folder for this in fawkes, yet and i assumed that there might be some reason behind that. If the `etc/scripts` folder was simply not needed as of yet, i can move the script of course :) no good reason on my side to keep it in the plugin folder (other than that this script is indeed very specific to the plugin itself).

TarikViehmann replied:
> Did you test this? I would have expected that `hello world` needs to be quoted, it's a string after all.
```
10:19:35.919432 FawkesMainThread: Disabling IPv6 support
10:19:37.814858 FawkesMainThread: Fawkes 1.3.0 startup complete
10:21:28.661057 SkillerSimulatorExecutionThread: BlackBoardNetworkHandler executes 'say{text="hello world"}'
10:21:28.663930 SkillerSimulatorExecutionThread: Executing 'say{text="hello world"}', will take 17.00 seconds
10:21:45.099255 SkillerSimulatorExecutionThread: BlackBoardNetworkHandler executes 'say{text="hello"}'
10:21:45.099282 SkillerSimulatorExecutionThread: Aborting execution of previous skill string 'say{text="hello world"}' for new goal
10:21:45.101129 SkillerSimulatorExecutionThread: Executing 'say{text="hello"}', will take 12.00 seconds
```

```
[tarikwork@localhost bin]$ ./skillet
Connection established
+# say{text="hello world"}
Executing: say{text="hello world"}
+# say{text="hello"}
Executing: say{text="hello"}
```

It works as you remove all `"` when parsing args: https://github.com/fawkesrobotics/fawkes/blob/edc998ce6b7eead0585b6ab2f29cd7047c4ceb0b/src/libs/execution_time_estimator/execution_time_estimator.cpp#L187-L193

morxa replied:
> > Is there a reason why you don't put the script into `etc/scripts`? This would also take care of installing it automatically.
> 
> I oriented myself on the plugin assisting tool of the `mongodb` (there it is indeed a c++ tool so it makes sense to only compile it if the plugin itself is built) plugin, so i initially placed it in the the plugin folder. Remembering our handling of scripts in fawkes-robotino i thought about placing it in `etc/scripts`, however there was no folder for this in fawkes, yet and i assumed that there might be some reason behind that. If the `etc/scripts` folder was simply not needed as of yet, i can move the script of course :) no good reason on my side to keep it in the plugin folder (other than that this script is indeed very specific to the plugin itself).

Yes, I think the reason is that we simply don't have any scripts yet. I think both ways are fine, I have a slight preference for putting it into `etc/scripts`.

That being said, the current script name is a bit too generic. Considering that this will be put in `bin`, the name should clearly convey what it does. What about `mongodb_skillsim_lookup.py`?

TarikViehmann replied:
> Yes, I think the reason is that we simply don't have any scripts yet. I think both ways are fine, I have a slight preference for putting it into `etc/scripts`.
> 
> That being said, the current script name is a bit too generic. Considering that this will be put in `bin`, the name should clearly convey what it does. What about `mongodb_skillsim_lookup.py`?

Yea the naming is poor atm, `mongodb_skillsim_lookup` sounds better. Putting it in `etc/scripts` won't install fawkes scripts when fawkes is used as submodule (the `etc/buildsys/root/linkscripts.mk` uses only `$(BASEDIR)` and hence it will not install scripts of fawkes). I guess for scripts it makes sense to install core scripts in any case, right?

morxa replied:
Yes, I guess that's a bug in `linkscripts.mk`.

TarikViehmann replied:
Fixed that in 158555237362d8de4316340e577a85bebd031189 and moved the script there in 9d7e944c29c93c7bf3de12f122544d21e64ed053.

TarikViehmann commented:
Thanks for the great feedback in the review!
